### PR TITLE
make restore fast by copying and comparing the lockfile.

### PR DESCRIPTION
**Pull request by matthid**
I think checking & validating this hash in our msbuild targets file should make building large paket based solutions a lot faster...

### CODE REVIEW

**Comment by matthid** [comment]
I'm really wondering that this in itself doesn't make it faster at all (3 instead of 5 secs).

From my first profiling attempt it looks like Argu is taking the most time :/

**Comment by matthid** [comment]
Also the magic mode bootstrapper is probably eating quite some time there

File: src/Paket/Program.fs [hunk 1]
```diff
@@ -745,9 +749,14 @@ let main() =
     use consoleTrace = Logging.event.Publish |> Observable.subscribe Logging.traceToConsole
     let paketVersion = AssemblyVersionInformation.AssemblyInformationalVersion
 
-
-
     try
+    let args = Environment.GetCommandLineArgs()
+    if args.Length = 1 && args.[0] = "restore" || args.Length = 2 && args.[0] = "--from-bootstrapper" && args.[1] = "restore" then
```
Comment (0x53A) [inline src/Paket/Program.fs]:
```F#
match args with
| [| "restore" |]
| [| "--from-bootstrapper" ; "restore" |]
```

?

Comment (matthid) [reply src/Paket/Program.fs]: yes looks better but might be slower :P 
The main problem is generating the sha512 hash in msbuild anyway ;) I don't think what I want is even possible...

**Comment by matthid** [comment]
Oh my god this seems to work

**Comment by matthid** [comment]
```
matth@DESKTOP-8VDEORF MINGW64 /c/proj/Paket (restore_speed)
$ PaketExePath=/c/proj/Paket/bin/paket.exe dotnet restore src/Paket.Core.preview3/Paket.Core.fsproj
  Paket version 5.91.0
  Performance:
   - Disk IO: 419 milliseconds
   - Runtime: 3 seconds
  Restore completed in 33.38 ms for C:\proj\Paket\src\Paket.Core.preview3\Paket.Core.fsproj.

matth@DESKTOP-8VDEORF MINGW64 /c/proj/Paket (restore_speed)
$ PaketExePath=/c/proj/Paket/bin/paket.exe dotnet restore src/Paket.Core.preview3/Paket.Core.fsproj
  Restore completed in 35.97 ms for C:\proj\Paket\src\Paket.Core.preview3\Paket.Core.fsproj.
```

File: src/Paket/Program.fs (hunk 1 above)
Comment (matthid) [reply src/Paket/Program.fs]: Ok fixed :)

**Comment by matthid** [comment]
IMHO this is good enough for now. It will hopefully help large netcore solutions and `Paket.VisualStudio`

**Comment by matthid** [comment]
WIP again, because netcore integration now needs to generate all files for all relevant frameworks beforehand

**Comment by matthid** [comment]
@alfonsogarciacaro I'm trying to change the restore process here a bit. In the future paket will write `"obj",projectFile.Name + "." + targetFramework + ".references"` instead of `"obj",projectFile.Name + ".references"`. I know that Fable uses this file.

We can still write the old file for compat. The question is what should we write into it? What is the target-framework fable uses? netstandard16?

**Comment by matthid** [comment]
I guess the build will be red because this is a bootstrapping problem :/

File: src/Paket.Core/PublicAPI.fs [hunk 2]
```diff
@@ -752,6 +736,42 @@ type Dependencies(dependenciesFileName: string) =
...
+
+        for nuspecFile in nuspecFileList do
+            if not (File.Exists nuspecFile) then
+                failwithf "Specified file '%s' does not exist." nuspecFile
+
+        let directDeps =
+            referencesFile.Groups
+            |> Seq.collect (fun kv -> kv.Value.NugetPackages)
+            |> Seq.map (fun i -> i.Name)
+            |> Set.ofSeq
```
Comment (matthid) [inline src/Paket.Core/PublicAPI.fs]: This might be a slight change in behavior. I just look into the references file and assume all things there are direct deps (which might not be true but I planned to add a warning when transitives are listed in the references file - should be issue open somewhere)

File: src/Paket.Core/PublicAPI.fs [hunk 3]
```diff
@@ -70,10 +70,7 @@ type Dependencies(dependenciesFileName: string) =
     static member Init() = Dependencies.Init(Directory.GetCurrentDirectory())
 
     /// Initialize paket.dependencies file in the given directory
-    static member Init(directory) =  Dependencies.Init(directory,false)
-
-    /// Initialize paket.dependencies file in the given directory
-    static member Init(directory,fromBootstrapper) =
```
Comment (matthid) [inline src/Paket.Core/PublicAPI.fs]: While technically a breaking change in the public API - I don't think anyone used these weird overloads.

**Comment by alfonsogarciacaro** [comment]
@matthid Fable 1.1.x uses `netstandard1.6`, I already publish a beta of Fable 1.2 which will use `netstandard2.0`, but I can make so the final Fable 1.2 looks for `"obj",projectFile.Name + "." + targetFramework + ".references"`. So, if we want to keep compatibility with Fable 1.1.x for a time, it should be fine to use `netstandard1.6`.

**Comment by matthid** [comment]
@alfonsogarciacaro Thanks will update. So I will also assume that if no netstandard16 is in the project file we no longer need that file.

**Comment by matthid** [comment]
TLDR:
 - This PR improves the speed of `paket restore` 
    - by saving "cache" files and quickly checking if they are dirty
    - by bypassing Argu (see https://github.com/fsprojects/Argu/issues/90)
 - This PR skips calling `paket restore` completely if we can detect in MSBuild that we are still up2date. (by using the same caching files as above)

**Comment by alfonsogarciacaro** [comment]
Awesome work, @matthid! Could you please check if this is going in the right direction? https://github.com/fable-compiler/Fable/commit/112fcd35e43d8092e1a586ec5cbb44c911b5aae2#diff-e47dbd37b262f2f0394b5704369d4b31R109

**Comment by matthid** [comment]
@alfonsogarciacaro Yes that looks exactly like it would do. Currently I'm thinking about renaming it to end with `.paket.references` but I haven't decided anything yet. Will let you know if I do :)

**Comment by matthid** [comment]
Ok green - finally :)

**Comment by matthid** [comment]
@Krzysztof-Cieslak Is Ionide calling paket restore directly or only `dotnet restore`? If the former we might want to take a look to benefit from this as well. I already updated https://github.com/fsprojects/Paket.VisualStudio/pull/140 to benefit as well.

**Comment by matthid** [comment]
Results in FAKE:

Before:
```
DotnetRestore                      00:05:35.9467550
DotnetPackage                      00:06:52.9450041
```
https://ci.appveyor.com/project/matthid/fake-6w516/build/1.0.106

After:
```
DotnetRestore                       0 <- no longer required
DotnetPackage                       00:04:32.9611943
```
https://ci.appveyor.com/project/matthid/fake-6w516/build/1.0.148

**Comment by alfonsogarciacaro** [comment]
Awesome! How is the suffix finally? `.references` or `.paket.references`? When will this be released?

**Comment by matthid** [comment]
@alfonsogarciacaro Yes I think it makes sense, will try adding it later today. And I think this can be released. But before that someone (@forki) needs to answer how the target files are updated. Because older target files should be compatible with newer `paket.exe` but definitely not the other way around (So we need to know if this can happen)

Once we know that we can make a "real" release with this.

**Comment by forki** [comment]
Paket install extracts the target files.

Am 28.08.2017 11:35 schrieb "Matthias Dittrich" <notifications@github.com>:

**Comment by matthid** [comment]
Perfect, thanks that should work. Will add another PR with the renaming later.
